src/Google.Logging.Log4Net/GoogleStackdriverAppender.cs

```diff
@@ -0,0 +1,274 @@
+    /// </remarks>
+    public partial class GoogleStackdriverAppender : AppenderSkeleton
+    {
+        // TODO:
+        // * Various argument validations
+        // * Send unsent logs on program exit? Using AppDomain.ProcessExit
+        //   - Note this only allows 2 seconds 
+
+        public const string s_logsLostWarningMessage = "Logs lost due to insufficient process-local storage: {0} -> {1}";
+        public const string s_lostDateTimeFmt = "yyyy-MM-dd' 'HH:mm:ss";
+
+        public GoogleStackdriverAppender() { }
+
+        // For testing only.
+        internal GoogleStackdriverAppender(LoggingServiceV2Client client,
+            IScheduler scheduler, IClock clock)
+        {
+            _client = GaxPreconditions.CheckNotNull(client, nameof(client));
+            _scheduler = GaxPreconditions.CheckNotNull(scheduler, nameof(scheduler));
+            _clock = GaxPreconditions.CheckNotNull(clock, nameof(clock));
+        }
+
+        private static readonly Dictionary<Level, LogSeverity> s_levelMap = new Dictionary<Level, LogSeverity>
+        {
+            // { Level.Off, LogSeverity.Info }, // Not for logging
+            // { Level.Log4Net_Debug, LogSeverity.Emergency }, // Duplicate of Emergency
+            { Level.Emergency, LogSeverity.Emergency },
+            { Level.Fatal, LogSeverity.Emergency },
+            { Level.Alert, LogSeverity.Alert },
+            { Level.Critical, LogSeverity.Critical },
+            { Level.Severe, LogSeverity.Critical },
+            { Level.Error, LogSeverity.Error },
+            { Level.Warn, LogSeverity.Warning },
+            { Level.Notice, LogSeverity.Notice },
+            { Level.Info, LogSeverity.Info },
+            { Level.Debug, LogSeverity.Debug },
+            // { Level.Fine, LogSeverity.Debug }, // Duplicate of Debug
+            { Level.Trace, LogSeverity.Debug },
+            // { Level.Finer, LogSeverity.Debug }, // Duplicate of Trace
+            { Level.Verbose, LogSeverity.Debug },
+            // { Level.Finest, LogSeverity.Debug }, // Duplicate of Verbose
+            //{ Level.All, LogSeverity.Debug }, // Not for logging
+        };
+
+        private object _enqueueLock = new object();
+        private string _instanceId = Guid.NewGuid().ToString();
+        private LoggingServiceV2Client _client;
+        private IScheduler _scheduler;
+        private IClock _clock;
+        private MonitoredResource _resource;
+        private string _logName;
+        private ILogQueue _logQ;
+        private LogUploader _logUploader;
+        private Task _initNextIdTask;
+        private long _nextId;
+
+        private List<Label> _customLabels = new List<Label>();
+        private HashSet<MetaDataType> _withMetaData = new HashSet<MetaDataType>();
+
+        public override void ActivateOptions()
+        {
+            base.ActivateOptions();
+
+            // Initialise services if not already initialised for testing
+            if (_client == null)
```

Comment:
What I really want is a null-safe assignment operator:
_client ?= LoggingServiceV2Client.Create();
But, alas, it doesn't exist.